salt/config/__init__.py

```diff
@@ -3284,6 +3292,42 @@ def _cache_id(minion_id, cache_file):
         log.error('Could not cache minion ID: {0}'.format(exc))
 
 
+def eval_minion_id_func(opts):
+    '''
+    Evaluate minion ID function if 'minion_id_type' is 'func'
+    and return the result
+    '''
+    if '__minion_id_func_evaluated' in opts:
+        return opts['id']
+
+    import salt.loader as loader
+
+    # split module and function and try loading the module
+    mod_fun = opts['id']
+    mod, fun = mod_fun.split('.')
+    if not opts.get('grains'):
+        # Get grains for use by the module
+        opts['grains'] = loader.grains(opts)
+
+    try:
+        minion_id_mod = loader.raw_mod(opts, mod, fun)
+        if not minion_id_mod:
+            raise KeyError
+        # we take whatever the module returns as the minion ID
+        newid = minion_id_mod[mod_fun]()
```

Comment:
No input interface from anywhere? There is a plenty of data in pillars, why not getting some?